File: CppGrammar/app/src/main/java/com/tao/myndktest/MainActivity.java
package com.tao.myndktest;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";

    // Used to load the 'native-lib' library on application startup.
    static {
        System.loadLibrary("native-lib");
    }

    private ListView listView;
    private List<ItemBean> dataList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Example of a call to a native method
        TextView tv = findViewById(R.id.sample_text);
        tv.setText(stringFromJNI());
        LogUtils.d(TAG,"onCreate --->");

        listView = findViewById(R.id.listView);
        initData();
    }

    private void initData() {
        dataList.add(new ItemBean("template模板", new Action() {
            @Override
            public void onAction() {
                new NativeHelper().testTemplate();
            }
        }));




        listView.setDivider(null);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                dataList.get(position).action.onAction();
            }
        });
        listView.setAdapter(new MyAdapter());

    }

    class MyAdapter extends BaseAdapter {

        @Override
        public int getCount() {
            return dataList.size();
        }

        @Override
        public Object getItem(int position) {
            return dataList.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            ViewHolder holder;
            if(convertView == null) {
                holder = new ViewHolder();
                convertView = LayoutInflater.from(MainActivity.this)
                        .inflate(R.layout.lv_item,null);
                holder.tv = convertView.findViewById(R.id.tv_item);
                convertView.setTag(holder);
            } else {
                holder = (ViewHolder) convertView.getTag();
            }
            String head = (position + 1) + ", ";
            holder.tv.setText( head + dataList.get(position).title);
            return convertView;
        }

        class ViewHolder {
            TextView tv;
        }
    }


    /**
     * A native method that is implemented by the 'native-lib' native library,
     * which is packaged with this application.
     */
    public native String stringFromJNI();


    public static int dip2px(Context context, float dpValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

}

File: FFmpegEnDe/app/src/main/java/com/tao/ffmpeg/MainActivity.kt
package com.tao.ffmpeg

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.core.app.ActivityCompat
import com.tao.ffmpeg.aac.FFmpegAACDecodeActivity
import com.tao.ffmpeg.aac.FFmpegAACEncodeActivity
import com.tao.ffmpeg.h264.FFmpegVideoEncoderActivity

class MainActivity : AppCompatActivity() {
    private val REQUEST_OK = 10001
    private val tag = "MainActivity"

    private val request_permission = arrayOf(
        Manifest.permission.WRITE_EXTERNAL_STORAGE,
        Manifest.permission.READ_EXTERNAL_STORAGE,
        Manifest.permission.RECORD_AUDIO,
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        LogUtils.d(tag, "onCreate checkPermission")
        if (!checkPermission(request_permission)) {
            ActivityCompat.requestPermissions(
                this, request_permission,REQUEST_OK
            )
        }
        // Example of a call to a native method
//        findViewById<TextView>(R.id.sample_text).text = stringFromJNI()
    }

    fun bt_aac_encode(view: View) {
        startActivity(Intent(this, FFmpegAACEncodeActivity::class.java))
    }

    fun bt_aac_decode(view: View) {
        startActivity(Intent(this, FFmpegAACDecodeActivity::class.java))
    }

    fun bt_h264_encode(view: View) {
        startActivity(Intent(this,FFmpegVideoEncoderActivity::class.java))
    }

    fun checkPermission(permissions: Array<String>): Boolean {
        for (permission in permissions) {
            if (ActivityCompat.checkSelfPermission(this, permission!!)
                != PackageManager.PERMISSION_GRANTED
            ) {
                return false
            }
        }
        LogUtils.d(tag, "checkPermission ok")
        return true
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String?>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        LogUtils.d(
            tag,
            "onRequestPermissionsResult: $requestCode"
        )
        if (requestCode != REQUEST_OK) {
            Toast.makeText(this, "权限申请失败", Toast.LENGTH_LONG).show()
        }
    }

    /**
     * A native method that is implemented by the 'native-lib' native library,
     * which is packaged with this application.
     */
//    external fun stringFromJNI(): String

    companion object {
        // Used to load the 'native-lib' library on application startup.
        init {
            System.loadLibrary("use_ffmpeg")
        }
    }


}